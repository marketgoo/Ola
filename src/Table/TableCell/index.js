import React from 'react'
import {default as PT} from 'prop-types'
import cx from 'classnames'
import { CaretDown, CaretUp } from '@phosphor-icons/react'

const TableCellOrder = ({ children, direction }) => {
  const icon = {
    asc: <CaretUp weight="fill" />,
    desc: <CaretDown weight="fill" />,
  }
  icon.null = icon.desc

  return (
    <>
      {children} {icon[direction]}
    </>
  )
}

const TableCell = ({ children, header, nowrap, selected, loading, orderable, direction, className, ...props }) => {
  return (
    <div
      className={cx('ola_table-cell', {
        'is-header': header,
        'is-nowrap': nowrap,
        'is-selected': selected,
        'is-orderable': orderable,
        'is-ordered': !!direction,
        'ola-skeleton is-loading': loading
      }, className)}
      role="cell"
      {...props}>
      <span>{
        loading
          ? null
          : orderable
            ? <TableCellOrder direction={direction}>{children}</TableCellOrder>
            : children
      }</span>
    </div>
  )
}

TableCell.defaultProps = {
  header: false,
  nowrap: false,
  selected: false,
  loading: false,
  orderable: false,
  direction: null,
  className: ''
}

TableCell.propTypes = {
  /** Indicates if the cell is loading */
  loading: PT.bool,
  /** Indicates if the cell is a header */
  header: PT.bool,
  /** Indicates if the text can be wrapped or not */
  nowrap: PT.bool,
  /** Indicates if the cell is selected */
  selected: PT.bool,
  /** Indicates if the cell is orderable */
  orderable: PT.bool,
  /** Order direction */
  direction: PT.oneOf(['asc', 'desc']),
  /** Extra className */
  className: PT.string,
  /** Child nodes */
  children: PT.oneOfType([
    PT.string,
    PT.arrayOf(PT.node),
    PT.node
  ]),
}

export default TableCell