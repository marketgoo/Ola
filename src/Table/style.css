.ola_table-container {
  display: block;
  width: 100%;
  height: auto;
  overflow: auto;

  &.is-sticky .ola_table {
    padding-top: var(--size-7);
  }
}

.ola_table {
  display: grid;
  grid-template-columns: var(--grid-template-columns);
  min-width: var(--min-width);
  max-height: var(--max-height);
  position: relative;
}

.ola_table-row {
  &.is-sticky {
    position: absolute;
    top: 0;
    left: 0;
    display: grid;
    z-index: 50;
  }
}

.ola_table-cell {
  display: inline-flex;
  align-items: center;
  font: var(--font-1-regular);
  border-bottom: 1px solid var(--color-neutral-200);
  height: var(--size-7);
  padding: 0 var(--size-2);
  background-color: var(--color-white);
  transition: all var(--duration);
  
  &.is-header {
    font: var(--font-1-bold);
    border-bottom: 1px solid var(--color-neutral-500);
  }

  &.is-nowrap,
  &.is-header {
    white-space: nowrap;
  }

  &:hover:not(.is-disabled), &.is-selected {
    background-color: var(--color-neutral-100);
  }

  &.is-disabled > * {
    opacity: .5;
    cursor: default;
    user-select: none;
    pointer-events: none;
  }
}