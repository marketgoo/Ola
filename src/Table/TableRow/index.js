import React from 'react'
import {default as PT} from 'prop-types'
import TableCell from '../TableCell'
import { uuidv4 } from '../../utils'

const TableRow = ({ children, header, sticky, stickyStyles, uuid }) => {
  const uuidGenerated = uuid || uuidv4()

  const cloneChildren = React.Children.map(children, child => {
    const config = { header }

    if (!header) {
      config['data-row'] = uuidGenerated
      config.onMouseEnter = () => {
        document.querySelectorAll(`[data-row="${uuidGenerated}"]`).forEach(el => {
          el.classList.add('is-hovered')
        })
      }
      config.onMouseLeave = () => {
        document.querySelectorAll(`[data-row="${uuidGenerated}"]`).forEach(el => {
          el.classList.remove('is-hovered')
        })
      }
    }

    return React.cloneElement(child, config)
  })

  if (sticky) {
    return (
      <div
        className="ola_table-row is-sticky"
        style={stickyStyles}
        role="row">
        {cloneChildren}
      </div>
    )
  }

  return (<>{cloneChildren}</>)
}

TableRow.defaultProps = {
  header: false,
  sticky: false,
  stickyStyles: {},
  uuid: null
}

TableRow.propTypes = {
  uuid: PT.string,
  sticky: PT.bool,
  stickyStyles: PT.object,
  /** Indicates if the row is a header row */
  header: PT.bool,
  /** Child nodes */
  children: PT.oneOfType([
    TableCell,
    PT.arrayOf(TableCell),
  ]),
}

export default TableRow