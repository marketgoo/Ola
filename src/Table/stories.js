import React, { useState } from 'react'
import Table from './'
import ButtonIcon from '../ButtonIcon'
import Check from '../Check'

const data = Array.from({ length: 12 }).map((_, index) => {
  return {
    id: index,
    title: `Page title test ${index}`,
    links: index * 2,
    popularity: index % 2 === 0 ? 'hight' : index % 3 === 0 ? 'medium' : 'low',
    checked: index % 3 === 0,
    action: <><ButtonIcon icon="success" color="positive" /><ButtonIcon icon="error" color="negative" /></>
  }
})

export default {
  title: 'Table/Table',
  component: Table,
  args: {
    stickyHeader: false,
    className: '',
    gridTemplateColumns: '',
    minWidth: '',
    maxHeight: '',
    loading: false
  },
  argTypes: {
    children: { control: { disable: true }}
  }
}

export const Base = (args) =>
  <Table {...args}>
    <Table.Header>
      <Table.Cell></Table.Cell>
      <Table.Cell nowrap>Title</Table.Cell>
      <Table.Cell>Links</Table.Cell>
      <Table.Cell>Popularity</Table.Cell>
      <Table.Cell>Action</Table.Cell>
    </Table.Header>
    {data.slice(0, 5).map((row, idx) => (
      <Table.Row key={idx}>
        <Table.Cell>
          <Check type="checkbox" name="foo" checked={row.checked} />
        </Table.Cell>
        <Table.Cell nowrap>{row.title}</Table.Cell>
        <Table.Cell>{row.links}</Table.Cell>
        <Table.Cell>{row.popularity}</Table.Cell>
        <Table.Cell>{row.action}</Table.Cell>
      </Table.Row>
    ))}
  </Table>

export const Loading = (args) =>
  <Table {...args}>
    <Table.Header>
      <Table.Cell></Table.Cell>
      <Table.Cell nowrap>Title</Table.Cell>
      <Table.Cell>Links</Table.Cell>
      <Table.Cell>Popularity</Table.Cell>
      <Table.Cell>Action</Table.Cell>
    </Table.Header>
    {data.slice(0, 5).map((row, idx) => (
      <Table.Row key={idx}>
        <Table.Cell>
          <Check type="checkbox" name="foo" checked={row.checked} />
        </Table.Cell>
        <Table.Cell nowrap>{row.title}</Table.Cell>
        <Table.Cell>{row.links}</Table.Cell>
        <Table.Cell>{row.popularity}</Table.Cell>
        <Table.Cell>{row.action}</Table.Cell>
      </Table.Row>
    ))}
  </Table>
Loading.args = {
  loading: true
}


export const CheckRowsExample = (args) => {
  const [selected, setSelected] = useState(data.filter(item => item.checked).map(item => item.id))
  const [allChecked, setAllChecked] = useState(false)

  const localData = data.slice(0, 5)

  return (
    <Table {...args}>
      <Table.Header>
        <Table.Cell>
          <Check type="checkbox" name="foo" checked={allChecked} onChange={
            (e) => {
              setAllChecked(e.target.checked)
              setSelected(e.target.checked ? data.map(item => item.id) : [])
            }
          } />
        </Table.Cell>
        <Table.Cell nowrap>Title</Table.Cell>
        <Table.Cell>Links</Table.Cell>
        <Table.Cell>Popularity</Table.Cell>
        <Table.Cell>Action</Table.Cell>
      </Table.Header>
      {localData.map((row, idx) => (
        <Table.Row key={idx}>
          <Table.Cell selected={selected.includes(row.id)}>
            <Check type="checkbox" name="foo" checked={selected.includes(row.id)} onChange={
              (e) => {
                if (e.target.checked) {
                  setSelected([...selected, row.id])
                } else {
                  setSelected(selected.filter(id => id !== row.id))
                }

                setAllChecked(selected.length === localData.length)
              }
            } />
          </Table.Cell>
          <Table.Cell selected={selected.includes(row.id)} nowrap>{row.title}</Table.Cell>
          <Table.Cell selected={selected.includes(row.id)}>{row.links}</Table.Cell>
          <Table.Cell selected={selected.includes(row.id)}>{row.popularity}</Table.Cell>
          <Table.Cell selected={selected.includes(row.id)}>{row.action}</Table.Cell>
        </Table.Row>
      ))}
    </Table>
  )
}
CheckRowsExample.args = {
  gridTemplateColumns: '60px 3fr 1fr 190px 2fr',
}

export const DisabledColumn = (args) =>
  <Table {...args}>
    <Table.Header>
      <Table.Cell></Table.Cell>
      <Table.Cell nowrap>Disabled Column</Table.Cell>
      <Table.Cell>Links</Table.Cell>
      <Table.Cell>Popularity</Table.Cell>
      <Table.Cell>Action</Table.Cell>
    </Table.Header>
    {data.slice(0, 5).map((row, idx) => (
      <Table.Row key={idx}>
        <Table.Cell><Check type="checkbox" name="foo" checked={row.checked} /></Table.Cell>
        <Table.Cell nowrap disabled>{row.title}</Table.Cell>
        <Table.Cell>{row.links}</Table.Cell>
        <Table.Cell>{row.popularity}</Table.Cell>
        <Table.Cell>{row.action}</Table.Cell>
      </Table.Row>
    ))}
  </Table>



export const SelectedColumn = (args) =>
  <Table {...args}>
    <Table.Header>
      <Table.Cell></Table.Cell>
      <Table.Cell nowrap>Title</Table.Cell>
      <Table.Cell>Selected Column</Table.Cell>
      <Table.Cell>Popularity</Table.Cell>
      <Table.Cell>Action</Table.Cell>
    </Table.Header>
    {data.slice(0, 5).map((row, idx) => (
      <Table.Row key={idx}>
        <Table.Cell><Check type="checkbox" name="foo" checked={row.checked} /></Table.Cell>
        <Table.Cell nowrap>{row.title}</Table.Cell>
        <Table.Cell selected>{row.links}</Table.Cell>
        <Table.Cell>{row.popularity}</Table.Cell>
        <Table.Cell>{row.action}</Table.Cell>
      </Table.Row>
    ))}
  </Table>

export const StickyHeader = (args) =>
  <Table {...args}>
    <Table.Header>
      <Table.Cell></Table.Cell>
      <Table.Cell nowrap>Title</Table.Cell>
      <Table.Cell>Links</Table.Cell>
      <Table.Cell>Popularity</Table.Cell>
      <Table.Cell>Action</Table.Cell>
    </Table.Header>
    {data.map((row, idx) => (
      <Table.Row key={idx}>
        <Table.Cell><Check type="checkbox" name="foo" checked={row.checked} /></Table.Cell>
        <Table.Cell nowrap>{row.title}</Table.Cell>
        <Table.Cell>{row.links}</Table.Cell>
        <Table.Cell>{row.popularity}</Table.Cell>
        <Table.Cell>{row.action}</Table.Cell>
      </Table.Row>
    ))}
  </Table>
StickyHeader.args = {
  maxHeight: '400px',
  stickyHeader: true
}

export const HorizontalScroll = (args) =>
  <Table {...args}>
    <Table.Header>
      <Table.Cell></Table.Cell>
      <Table.Cell nowrap>Title</Table.Cell>
      <Table.Cell>Links</Table.Cell>
      <Table.Cell>Popularity</Table.Cell>
      <Table.Cell>Action</Table.Cell>
    </Table.Header>
    {data.slice(0, 5).map((row, idx) => (
      <Table.Row key={idx}>
        <Table.Cell><Check type="checkbox" name="foo" checked={row.checked} /></Table.Cell>
        <Table.Cell nowrap>{row.title}</Table.Cell>
        <Table.Cell>{row.links}</Table.Cell>
        <Table.Cell>{row.popularity}</Table.Cell>
        <Table.Cell>{row.action}</Table.Cell>
      </Table.Row>
    ))}
  </Table>
HorizontalScroll.args = {
  minWidth: '2000px',
}

export const StickyHeaderHorizontalScroll = (args) =>
  <Table {...args}>
    <Table.Header>
      <Table.Cell></Table.Cell>
      <Table.Cell nowrap>Title</Table.Cell>
      <Table.Cell>Links</Table.Cell>
      <Table.Cell>Popularity</Table.Cell>
      <Table.Cell>Action</Table.Cell>
    </Table.Header>
    {data.map((row, idx) => (
      <Table.Row key={idx}>
        <Table.Cell><Check type="checkbox" name="foo" checked={row.checked} /></Table.Cell>
        <Table.Cell nowrap>{row.title}</Table.Cell>
        <Table.Cell>{row.links}</Table.Cell>
        <Table.Cell>{row.popularity}</Table.Cell>
        <Table.Cell>{row.action}</Table.Cell>
      </Table.Row>
    ))}
  </Table>
StickyHeaderHorizontalScroll.args = {
  minWidth: '2000px',
  maxHeight: '400px',
  stickyHeader: true
}

export const CustomGridTemplate = (args) =>
  <>
    <p><strong><code>gridTemplateColumns: &quot;60px 3fr 1fr 190px 2fr&quot;</code></strong></p>
    <Table {...args}>
      <Table.Header>
        <Table.Cell>60px</Table.Cell>
        <Table.Cell nowrap>3fr</Table.Cell>
        <Table.Cell>1fr</Table.Cell>
        <Table.Cell>190px</Table.Cell>
        <Table.Cell>2fr</Table.Cell>
      </Table.Header>
      {data.slice(0, 5).map((row, idx) => (
        <Table.Row key={idx}>
          <Table.Cell><Check type="checkbox" name="foo" checked={row.checked} /></Table.Cell>
          <Table.Cell nowrap>{row.title}</Table.Cell>
          <Table.Cell>{row.links}</Table.Cell>
          <Table.Cell>{row.popularity}</Table.Cell>
          <Table.Cell>{row.action}</Table.Cell>
        </Table.Row>
      ))}
    </Table>
  </>
CustomGridTemplate.args = {
  gridTemplateColumns: '60px 3fr 1fr 190px 2fr',
}
