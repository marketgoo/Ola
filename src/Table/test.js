import React from 'react'
import Table from './'
import renderer from 'react-test-renderer'


it('Default Table', () => {
  const tree = renderer
    .create(
      <Table>
        <Table.Row>
          <Table.Cell></Table.Cell>
          <Table.Cell nowrap>Title</Table.Cell>
          <Table.Cell>Links</Table.Cell>
          <Table.Cell>Popularity</Table.Cell>
          <Table.Cell>Action</Table.Cell>
        </Table.Row>
        <Table.Row>
          <Table.Cell>
            Check
          </Table.Cell>
          <Table.Cell nowrap>The title</Table.Cell>
          <Table.Cell>The links</Table.Cell>
          <Table.Cell>The popularity</Table.Cell>
          <Table.Cell>The actions</Table.Cell>
        </Table.Row>
      </Table>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
