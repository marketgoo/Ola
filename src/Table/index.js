import React, { useEffect, useRef } from 'react'
import {default as PT} from 'prop-types'
import cx from 'classnames'
import TableRow from './TableRow'
import TableCell from './TableCell'
import TableHeader from './TableHeader'
import { debounce } from '../utils'


const Table = ({ stickyHeader, minWidth, maxHeight, className, gridTemplateColumns, loading, children }) => {
  const [stickyStyles, setStickyStyles] = React.useState({})
  const tableRef = useRef()
  let numColumns = 0

  const cloneChildren = React.Children.map(children, (child) => {
    if (child.props.__type__ === 'TableHeader') {
      numColumns = React.Children.count(child.props.children)
      return React.cloneElement(child, { sticky: stickyHeader, stickyStyles })
    }

    let childrenRow = child.props.children

    if (loading) {
      childrenRow = React.Children.map(child.props.children, (child) => {
        return React.cloneElement(child, { loading: true })
      })
    }

    return React.cloneElement(child, { sticky: false, children: childrenRow, header: false })
  })

  const handleScrollResize = debounce((e) => {
    const eventType = e?.nativeEvent ? e?.nativeEvent.type : e?.type

    const tableStyles = window.getComputedStyle(tableRef.current)
    const nStickyStyles = {
      ...stickyStyles,
      top: e ? `${e.target.scrollTop}px` : '0'
    }

    if (!eventType || eventType === 'resize') {
      const firstElements = Array.from(tableRef.current.querySelectorAll('.ola_table-cell:not(.is-header)')).slice(0, numColumns)
      nStickyStyles.width = tableStyles.width
      nStickyStyles.gridTemplateColumns = firstElements.map(elem => `${elem.offsetWidth}px`).join(' ')
    }

    setStickyStyles(nStickyStyles)
  })

  useEffect(() => {
    if (stickyHeader) {
      handleScrollResize()

      window.addEventListener('resize', handleScrollResize)
    }

    return () => {
      window.removeEventListener('resize', handleScrollResize)
    }
  }, [])

  return (
    <div className={
      cx(
        'ola_table-container',
        { 'is-sticky': stickyHeader },
        className)
    } onScroll={stickyHeader ? handleScrollResize : null}>
      <div ref={tableRef} className='ola_table' role='table' style={{
        '--min-width': minWidth,
        '--max-height': maxHeight,
        '--grid-template-columns': gridTemplateColumns || `repeat(${numColumns}, 1fr)`
      }}>
        {cloneChildren}
      </div>
    </div>
  )
}

Table.Cell = TableCell
Table.Row = TableRow
Table.Header = TableHeader

Table.defaultProps = {
  className: '',
  stickyHeader: false,
  minWidth: '100%',
  maxHeight: 'unset',
  gridTemplateColumns: null,
  loading: false
}

Table.propTypes = {
  /**
   * To overwrite the grid-template-columns. The default value is `repeat(NUM_COLUMNS, 1fr)`
   * Accept any valid CSS [grid-template-columns](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-columns) value. */
  gridTemplateColumns: PT.string,
  /** Set the status table as loading */
  loading: PT.bool,
  /** Minimum width of the table in css valid value */
  minWidth: PT.string,
  /** Maximun height of the table in css valid value */
  maxHeight: PT.string,
  /** Table header is sticky */
  stickyHeader: PT.bool,
  /** Extra className */
  className: PT.string,
  /** Children nodes. */
  children: PT.oneOfType([
    TableRow,
    PT.arrayOf(TableRow),
    TableHeader,
    PT.arrayOf(TableHeader),
  ]).isRequired,
}

export default Table