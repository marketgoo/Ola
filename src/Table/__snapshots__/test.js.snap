// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Table 1`] = `
<div
  className="ola_table-container"
  onScroll={null}
>
  <div
    className="ola_table"
    role="table"
    style={
      Object {
        "--grid-template-columns": "repeat(5, 1fr)",
        "--max-height": "unset",
        "--min-width": "100%",
      }
    }
  >
    <div
      className="ola_table-cell is-align--left"
      data-row="aaa"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="cell"
    >
      <span />
    </div>
    <div
      className="ola_table-cell is-align--left is-nowrap"
      data-row="aaa"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="cell"
    >
      <span>
        Title
      </span>
    </div>
    <div
      className="ola_table-cell is-align--left"
      data-row="aaa"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="cell"
    >
      <span>
        Links
      </span>
    </div>
    <div
      className="ola_table-cell is-align--left"
      data-row="aaa"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="cell"
    >
      <span>
        Popularity
      </span>
    </div>
    <div
      className="ola_table-cell is-align--left"
      data-row="aaa"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="cell"
    >
      <span>
        Action
      </span>
    </div>
    <div
      className="ola_table-cell is-align--left"
      data-row="bbb"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="cell"
    >
      <span>
        Check
      </span>
    </div>
    <div
      className="ola_table-cell is-align--left is-nowrap"
      data-row="bbb"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="cell"
    >
      <span>
        The title
      </span>
    </div>
    <div
      className="ola_table-cell is-align--left"
      data-row="bbb"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="cell"
    >
      <span>
        The links
      </span>
    </div>
    <div
      className="ola_table-cell is-align--left"
      data-row="bbb"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="cell"
    >
      <span>
        The popularity
      </span>
    </div>
    <div
      className="ola_table-cell is-align--left"
      data-row="bbb"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="cell"
    >
      <span>
        The actions
      </span>
    </div>
  </div>
</div>
`;
